compiler/damlc/daml-doc/src/DA/Daml/Doc/Transform.hs

```diff
@@ -8,61 +8,17 @@ module DA.Daml.Doc.Transform
   ) where
 
 import DA.Daml.Doc.Types
-import DA.Daml.Doc.Annotate
+import DA.Daml.Doc.Transform.Options
+import DA.Daml.Doc.Transform.Annotations
+import DA.Daml.Doc.Transform.Instances
+import DA.Daml.Doc.Transform.DropEmpty
 
 import Data.Maybe
-import Data.List.Extra
-import System.FilePath (pathSeparator) -- because FilePattern uses it
-import System.FilePattern
-import qualified Data.Text as T
-import qualified Data.Map as Map
-import qualified Data.Set as Set
-
-data TransformOptions = TransformOptions
-    { to_includeModules :: Maybe [String]
-    , to_excludeModules :: Maybe [String]
-    , to_excludeInstances :: Set.Set String
-    , to_dataOnly :: Bool -- ^ do not generate docs for functions and classes
-    , to_ignoreAnnotations :: Bool -- ^ ignore MOVE and HIDE annotations
-    , to_omitEmpty :: Bool -- ^ omit all items that do not have documentation
-    }
-
-defaultTransformOptions :: TransformOptions
-defaultTransformOptions = TransformOptions
-    { to_includeModules = Nothing
-    , to_excludeModules = Nothing
-    , to_excludeInstances = Set.empty
-    , to_dataOnly = False
-    , to_ignoreAnnotations = False
-    , to_omitEmpty = False
-    }
-
-keepModule :: TransformOptions -> ModuleDoc -> Bool
-keepModule TransformOptions{..} m = includeModuleFilter && excludeModuleFilter
-  where
-    includeModuleFilter :: Bool
-    includeModuleFilter = maybe True moduleMatchesAny to_includeModules
-
-    excludeModuleFilter :: Bool
-    excludeModuleFilter = maybe True (not . moduleMatchesAny) to_excludeModules
-
-    moduleMatchesAny :: [String] -> Bool
-    moduleMatchesAny ps = any (?== name) (map withSlashes ps)
-
-    withSlashes :: String -> String
-    withSlashes = replace "." [pathSeparator]
-
-    name :: String
-    name = withSlashes . T.unpack . unModulename . md_name $ m
-
-keepInstance :: TransformOptions -> InstanceDoc -> Bool
-keepInstance TransformOptions{..} InstanceDoc{..} =
-    let nameM = T.unpack . unTypename <$> getTypeAppName id_type
-    in maybe True (`Set.notMember` to_excludeInstances) nameM
 
 applyTransform :: TransformOptions -> [ModuleDoc] -> [ModuleDoc]
 applyTransform opts@TransformOptions{..}
-    = distributeInstanceDocs opts
+    = (if to_dropOrphanInstances then map pruneOrphanInstances else id)
```

Comment:
This looks like you are erasing all instances. I would have expected that we keep non-orphan instances? And even for orphan instances, I would have expected that we want to keep them and only remove whatever indicates that they are orphan.

Reply:
OK, I think I understand this now. The instances are attached to the datatypes in `distributeInstanceDocs`. Afterwards, this field is only used to display a module-level list of orphan instances so we don’t care about it and can just erase it.

So the previous behavior was:
1. An orphan instance is attached to the datatype in `distributeInstanceDocs`.
2. In addition to that the instance is listed in the list of orphan instances in the module.

The new behavior (with the flag turned on) is:

1. The orphan instance is attached to the datatype.
2. The list of orphan instances will be empty.

Might be nice to add a comment here that explains why we can erase this field completely.

Reply:
I agree that some docs here would be good :-)

Reply:
I ended up rewriting the filter to only drop instances that were marked as orphans, since that makes the most sense to me after thinking about it some more. (Also I moved this filter to DA.Daml.Doc.Transform.Instances). Maybe I will revisit this later, but at least this way it's clear what "drop orphan instances" actually does.

Reply:
Sounds good, thank you!